python/torch_mlir/dialects/torch/importer/jit_ir/build_tools/library_generator.py

```diff
@@ -0,0 +1,174 @@
+    # (Scalar, Scalar) -> (Tensor). However, once a module gets
+    # jit.scripted, all math ops that work on scalars becomes
+    # (Scalar, Scalar) -> (Scalar) ops. So to get the right result
+    # dtype, we use the tensor-tensor promotion rules.
+    return _promote_tensor_tensor(0, lhs_dtype, 0, rhs_dtype)
+
+def promote_dtypes(ranks: List[Optional[int]],
+                   dtypes: List[torch.dtype]) -> torch.dtype:
+    return __torch_mlir_internal_promote_dtypes(ranks, dtypes)
+
+def __torch_mlir_internal_promote_dtypes(ranks: List[Optional[int]],
+                                         dtypes: List[torch.dtype]
+                                         ) -> torch.dtype:
+    lhs_optional_rank = ranks[0]
+    lhs_dtype = dtypes[0]
+    for rhs_optional_rank, rhs_dtype in zip(ranks, dtypes):
+        if lhs_optional_rank is None and rhs_optional_rank is None:
+            lhs_dtype = _promote_scalar_scalar(lhs_dtype, rhs_dtype)
+        elif lhs_optional_rank is None and rhs_optional_rank is not None:
+            lhs_dtype = _promote_scalar_tensor(
+                lhs_dtype, rhs_optional_rank, rhs_dtype)
+            lhs_optional_rank = rhs_optional_rank
+        elif lhs_optional_rank is not None and rhs_optional_rank is None:
+            lhs_dtype = _promote_scalar_tensor(
+                rhs_dtype, lhs_optional_rank, lhs_dtype)
+        elif lhs_optional_rank is not None and rhs_optional_rank is not None:
+            lhs_dtype = _promote_tensor_tensor(
+                lhs_optional_rank, lhs_dtype, rhs_optional_rank, rhs_dtype)
+            lhs_optional_rank = max(lhs_optional_rank, rhs_optional_rank)
+    return lhs_dtype
+
+def not_present_in_registry(f):
+    """Decorator for shape functions not present in the shape registry.
+
+    This can happen for "valsem" ops that we have in Torch-MLIR, such as
+    torch.valsem.aten.fill.Scalar, which are consistent with PyTorch conventions
+    (e.g. being the value-semantic correspondent of torch.aten.fill_.Scalar),
+    but that for whatever reason are not present in PyTorch. Such ops are useful
+    to keep certain passes within Torch-MLIR as consistent as possible.
+    For such ops, in the shape library generator, we treat them as if they
+    were registered torch ops (so we don't put "valsem" on them), to keep the
+    generator consistent.
+
+    To check if this decorator has been applied, use
+    `hasattr(f, "_not_present_in_registry")`.
+    """
+    f._not_present_in_registry = None
+    return f
+
+def _verify_signature_matches_registry(f, registry: Registry):
+    source = inspect.getsource(f)
+    signature = None
+    for line in source.splitlines():
+        if line.startswith("def "):
+            signature = line
+            break
+    assert signature is not None, f"Could not find signature for {f.__name__}"
+    assert "〡" in signature, f"Malformed signature {signature}. Signature missing the character `〡`"
+    f_name, overload = f.__name__.split("〡")
```

Comment:
overload has another meaning (it is the 3rd atom in f_name). So perhaps function_kind could be better